eslint-local-rules.js

```diff
@@ -0,0 +1,204 @@
+  // those nodes should not have a side effect when evaluated.
+  //
+  // This list is probably not complete, feel free to add more cases if you encounter an unhandled
+  // node.
+  switch (node.type) {
+    case 'Program':
+      node.body.forEach((child) => reportPotentialSideEffect(context, child))
+      return
+    case 'TemplateLiteral':
+      node.expressions.forEach((child) => reportPotentialSideEffect(context, child))
+      return
+    case 'ExportNamedDeclaration':
+    case 'ExportAllDeclaration':
+    case 'ImportDeclaration':
+      if (node.declaration) {
+        reportPotentialSideEffect(context, node.declaration)
+      } else if (
+        node.source &&
+        node.importKind !== 'type' &&
+        !isAllowedImport(context.getFilename(), node.source.value)
+      ) {
+        context.report({
+          node: node.source,
+          message: 'This file cannot import modules with side-effects',
+        })
+      }
+      return
+    case 'VariableDeclaration':
+      node.declarations.forEach((child) => reportPotentialSideEffect(context, child))
+      return
+    case 'VariableDeclarator':
+      if (node.init) {
+        reportPotentialSideEffect(context, node.init)
+      }
+      return
+    case 'ArrayExpression':
+      node.elements.forEach((child) => reportPotentialSideEffect(context, child))
+      return
+    case 'UnaryExpression':
+      reportPotentialSideEffect(context, node.argument)
+      return
+    case 'ObjectExpression':
+      node.properties.forEach((child) => reportPotentialSideEffect(context, child))
+      return
+    case 'SpreadElement':
+      reportPotentialSideEffect(context, node.argument)
+      return
+    case 'Property':
+      reportPotentialSideEffect(context, node.key)
+      reportPotentialSideEffect(context, node.value)
+      return
+    case 'BinaryExpression':
+      reportPotentialSideEffect(context, node.left)
+      reportPotentialSideEffect(context, node.right)
+      return
+    case 'TSAsExpression':
+    case 'ExpressionStatement':
+      reportPotentialSideEffect(context, node.expression)
+      return
+    case 'MemberExpression':
+      reportPotentialSideEffect(context, node.object)
+      reportPotentialSideEffect(context, node.property)
+      return
+    case 'FunctionExpression':
+    case 'ArrowFunctionExpression':
+    case 'FunctionDeclaration':
+    case 'ClassDeclaration':
+    case 'TSEnumDeclaration':
+    case 'TSInterfaceDeclaration':
+    case 'TSTypeAliasDeclaration':
+    case 'TSDeclareFunction':
+    case 'Literal':
+    case 'Identifier':
+      return
+    case 'CallExpression':
+      if (isAllowedCallExpression(node)) {
+        return
+      }
+    case 'NewExpression':
+      if (isAllowedNewExpression(node)) {
+        return
+      }
+  }
+
+  // If the node doesn't match any of the condition above, report it
+  context.report({ node, message: `${node.type} not allowed in types, constants and internal files` })
```

Comment:
the message does not seem right anymore, maybe saying that it is a possible side effect and we should consider to do something else